### PR TITLE
Global patching for scikit-learn

### PR DESCRIPTION
**Pull request by Pahandrovich**

### CODE REVIEW

**Comment by PetrovKP** [comment]
```python
python -m sklearnex patch
```
and 
```python
python -m sklearnex unpatch
```
right?

**Comment by Pahandrovich** [comment]
@PetrovKP, yes

**Comment by Pahandrovich** [comment]
other possible option:
`python -m sklearnex.global patch/unpatch`

and

`python -m sklearnex.global_patch`
`python -m sklearnex.global_unpatch`

File: sklearnex/global_patching.py [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+    return \
+"""try:
+    from sklearnex import patch_sklearn
+    patch_sklearn()
+    del patch_sklearn
+except (ImportError, ModuleNotFoundError):
+    pass"""
+
+
+def global_patching():
```
Comment (SmirnovEgorRu) [inline sklearnex/global_patching.py:31]: ```patch_sklearn_global```?
or
```patch_sklearn(global=True)```?

**Comment by Pahandrovich** [comment]
patch line:
python -m sklearnex.global patch_sklearn -a algo_1 algo_2 algo_2 -nv

File: sklearnex/_utils.py [hunk 2]
```diff
@@ -63,3 +63,69 @@ def get_patch_message(s):
...
+    init_file_path = sklearn.__file__
+    distributor_file_path = os.path.join(os.path.dirname(init_file_path),
+                                         "_distributor_init.py")
+
+    with open(distributor_file_path, 'r', encoding='utf-8') as distributor_file:
+        lines = distributor_file.read()
+        if re.search(get_patch_str_re(), lines):
+            lines = re.sub(get_patch_str_re(), '', lines)
+
+    with open(distributor_file_path, 'w', encoding='utf-8') as distributor_file:
```
Comment (PetrovKP) [inline sklearnex/_utils.py:103]: Does this mean it will always overwrite?

File: sklearnex/_utils.py [hunk 3]
```diff
@@ -63,3 +63,69 @@ def get_patch_message(s):
 def get_sklearnex_version(rule):
     from daal4py.sklearn._utils import daal_check_version
     return daal_check_version(rule)
+
+
+def get_patch_str(name=None, verbose=True):
```
Comment (PetrovKP) [inline sklearnex/_utils.py:68]: It's need when you have `get_patch_str_re`?

File: sklearnex/_utils.py [hunk 4]
```diff
@@ -63,3 +63,69 @@ def get_patch_message(s):
...
     from daal4py.sklearn._utils import daal_check_version
     return daal_check_version(rule)
+
+
+def get_patch_str(name=None, verbose=True):
+    return f"""try:
+    from sklearnex import patch_sklearn
+    patch_sklearn(name={str(name)}, verbose={str(verbose)})
+    del patch_sklearn
+except (ImportError, ModuleNotFoundError):
```
Comment (PetrovKP) [inline sklearnex/_utils.py:73]: `ModuleNotFoundError` inherited from `ImportError`

File: sklearnex/dispatcher.py [hunk 5]
```diff
@@ -52,11 +52,16 @@ def get_patch_names():
     return list(get_patch_map().keys())
 
 
-def patch_sklearn(name=None, verbose=True):
+def patch_sklearn(name=None, verbose=True, _global=False):
```
Comment (PetrovKP) [inline sklearnex/dispatcher.py:55]: bad, strong interface change. move common code, for example:
```python
def patch_sklearn(name=None, verbose=True):
   detail_patch_sklearn(name=name, verbose=verbose, global=False)

def detail_patch_sklearn(name, verbose, global):
   ....
```

File: sklearnex/_utils.py (hunk 3 above)
Comment (PetrovKP) [inline sklearnex/_utils.py:68]: I think functional for global patch move to global folder

File: sklearnex/global/__main__.py [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+    if args.action == "patch_sklearn":
+        patch_sklearn(name=args.algorithm, verbose=(not args.no_verbose), _global=True)
+        return
+
+    if args.action == "unpatch_sklearn":
+        unpatch_sklearn(_global=True)
+        return
+
+
+sys.exit(_main())
```
Comment (PetrovKP) [inline sklearnex/global/__main__.py:60]: It's need?

File: sklearnex/global/__main__.py [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+                        help="Additional info about patching")
+    parser.add_argument('--algorithm', '-a', action='extend', type=str, nargs="+",
+                        help="Name of algorithm to global patch")
+    args = parser.parse_args()
+
+    if args.action == "patch_sklearn":
+        patch_sklearn(name=args.algorithm, verbose=(not args.no_verbose), _global=True)
+        return
+
+    if args.action == "unpatch_sklearn":
```
Comment (PetrovKP) [inline sklearnex/global/__main__.py:55]:
```suggestion
    elif args.action == "unpatch_sklearn":

```

File: .ci/pipeline/ci.yml [hunk 8]
```diff
@@ -134,16 +134,36 @@ jobs:
...
-    displayName: conda build
+    displayName: install sklearnex
+  - script: |
+      . /usr/share/miniconda/etc/profile.d/conda.sh
+      conda activate CB
+      cd ..
+      ret_code=0
+      python -m sklearnex.global patch_sklearn -a svc
+      ret_code=$(($ret_code + $?))
+      python -c "from sklearn.svm import SVC; assert SVC.__module__.startswith('daal4py') or SVC.__module__.startswith('sklearnex')"
```
[the remark this answers is not in the document]
Comment (Pahandrovich) [reply .ci/pipeline/ci.yml:155]: yes, I'll create .sh and .bat files

File: sklearnex/_utils.py (hunk 3 above)
Comment (Pahandrovich) [reply sklearnex/_utils.py:68]: get_patch_str need for write right patch lines in _distributor_init file
get_patch_str_re need to search our changes in _distributor_init

Comment (Pahandrovich) [reply sklearnex/_utils.py:68]: agree, thanks

File: sklearnex/_utils.py (hunk 2 above)
Comment (Pahandrovich) [reply sklearnex/_utils.py:103]: Yes, we overwrite file with patching string

File: sklearnex/dispatcher.py (hunk 5 above)
Comment (Pahandrovich) [reply sklearnex/dispatcher.py:55]: agree, thanks)

File: sklearnex/glob/__main__.py [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+    class ExtendAction(argparse.Action):
+        def __call__(self, parser, namespace, values, option_string=None):
+            items = getattr(namespace, self.dest) or []
+            items.extend(values)
+            setattr(namespace, self.dest, items)
+
+    parser = argparse.ArgumentParser(
+        prog="python -m sklearnex.glob",
+        description="""
+            Patch your all Scikit-learn applications using Intel(R) Extension for
```
Comment (PetrovKP) [inline sklearnex/glob/__main__.py:35]: @outoftardis can you please review text?

File: sklearnex/glob/dispatcher.py [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+                                         "_distributor_init.py")
+
+    with open(distributor_file_path, 'r', encoding='utf-8') as distributor_file:
+        lines = distributor_file.read()
+        if re.search(get_patch_str_re(), lines):
+            lines = re.sub(get_patch_str_re(), '', lines)
+
+    with open(distributor_file_path, 'w', encoding='utf-8') as distributor_file:
+        distributor_file.write(lines + "\n" + get_patch_str(name, verbose) + "\n")
+        print("Scikit-learn successfully patched by Intel(R) Extension for Scikit-learn")
```
Comment (Pahandrovich) [inline sklearnex/glob/dispatcher.py:55]: @outoftardis, can you review text, please?
and 65, 74, 80 lines

File: sklearnex/glob/__main__.py (hunk 9 above)
Comment (outoftardis) [reply sklearnex/glob/__main__.py:35]:
```suggestion
            Patch all your Scikit-learn applications using Intel(R) Extension for
```

File: sklearnex/glob/__main__.py [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+            scikit-learn, optimizing solvers of
+            scikit-learn with Intel(R) oneAPI Data Analytics Library.
```
Comment (outoftardis) [inline sklearnex/glob/__main__.py:37]: unclear to me what you are saying with the second half of the sentence, and I'm also not sure if you really need that info in this description in the first place

File: sklearnex/glob/__main__.py [hunk 12]
```diff
@@ -0,0 +1,60 @@
...
+            """,
+        formatter_class=argparse.ArgumentDefaultsHelpFormatter)
+
+    parser.register('action', 'extend', ExtendAction)
+    parser.add_argument('action', choices=["patch_sklearn", "unpatch_sklearn"],
+                        help="Enable or Disable patching")
+    parser.add_argument('--no-verbose', '-nv', action='store_false',
+                        help="Additional info about patching")
```
Comment (outoftardis) [inline sklearnex/glob/__main__.py:45]:
```suggestion
                        help="Additional information about the patching")
```

what is the action here? e.g. in line 43 the action is enable/disable, you need a verb here as well.
plus I'd assume that no-verbose means that you are turning off the verbose mode, meaning you are not providing any additional info

File: sklearnex/glob/__main__.py [hunk 13]
```diff
@@ -0,0 +1,60 @@
...
+    parser.add_argument('--algorithm', '-a', action='extend', type=str, nargs="+",
+                        help="Name of algorithm to global patch")
```
Comment (outoftardis) [inline sklearnex/glob/__main__.py:47]:
```suggestion
                        help="Provide the name of an algorithm to be patched globally")
```

File: sklearnex/glob/dispatcher.py [hunk 14]
```diff
@@ -0,0 +1,80 @@
...
+    with open(distributor_file_path, 'r', encoding='utf-8') as distributor_file:
+        lines = distributor_file.read()
+        if not re.search(get_patch_str_re(), lines):
+            print("Scikit-learn didn't patched. Nothing to unpatch\n")
+            return
+        lines = re.sub(get_patch_str_re(), '', lines)
+
+    with open(distributor_file_path, 'w', encoding='utf-8') as distributor_file:
+        distributor_file.write(lines)
+        print("Scikit-learn has been successfully globally unpatched")
```
Comment (outoftardis) [inline sklearnex/glob/dispatcher.py:80]:
```suggestion
        print("Scikit-learn was successfully globally unpatched")
```

File: sklearnex/glob/dispatcher.py (hunk 10 above)
Comment (Pahandrovich) [reply sklearnex/glob/dispatcher.py]: Fixed

**Review by PetrovKP — APPROVED** [review]

**Pull request merged**